### PR TITLE
*: Remove leading comment when check IsQuery

### PR DESCRIPTION
**Pull request by shenli**
IsQuery check the prefix of sql text, so we need to remove comments.

### CODE REVIEW

File: tidb.go [hunk 1]
```diff
@@ -243,9 +243,24 @@ func NewStore(uri string) (kv.Storage, error) {
 
 var queryStmtTable = []string{"explain", "select", "show", "execute", "describe", "desc"}
 
+func trimSQL(sql string) string {
```
Comment (ngaut) [inline tidb.go]: s/trim/stripComments  ?

Comment (ngaut) [inline tidb.go]: And comments may start with --

Comment (shenli) [inline tidb.go]: If comments start with -- or # the entire line is comment. Only /**/ can mix with sql statement.

Comment (shenli) [inline tidb.go]: trim comment and space

File: parser/parser_test.go [hunk 2]
```diff
@@ -335,19 +335,21 @@ func (s *testParserSuite) TestParser0(c *C) {
 
 		// For select with where clause
 		{"SELECT * FROM t WHERE 1 = 1", true},
+
+		// For comment in query
+		{"/*comment*/ /*comment*/ select c /* this is a comment */ from t;", true},
```
Comment (qiuyesuifeng) [inline parser/parser_test.go]: It is better to add different types of comments.
like 
`/* xxx */ select * from t`
`select * /* xxx */ from t;`
`select * from t /* xxx */`

File: parser/parser_test.go [hunk 3]
```diff
@@ -335,19 +335,21 @@ func (s *testParserSuite) TestParser0(c *C) {
...
 	for _, t := range table {
-		fmt.Printf("%s\n", t.src)
 		l := NewLexer(t.src)
 		ok := yyParse(l) == 0
 		c.Assert(ok, Equals, t.ok, Commentf("source %v", t.src))
 
 		switch ok {
 		case true:
-			c.Assert(len(l.errs), Equals, 0)
+			c.Assert(len(l.errs), Equals, 0, Commentf("src: %s", t.src))
```
Comment (qiuyesuifeng) [inline parser/parser_test.go]: Can use HasLen instead of old code.

File: tidb.go (hunk 1 above)
Comment (ngaut) [inline tidb.go]: We need to strip comments start with --
see https://github.com/pingcap/tidb/issues/199

Comment (siddontang) [inline tidb.go]: Please add test case.

Btw, only trimming prefix is ok?

Comment (shenli) [inline tidb.go]: There are some test cases in tidb_test.go

File: tidb.go [hunk 4]
```diff
@@ -243,9 +243,24 @@ func NewStore(uri string) (kv.Storage, error) {
...
+	// Trim space.
+	sql = strings.TrimSpace(sql)
+	// Trim leading /*comment*/
+	// There may be multiple comments
+	for strings.HasPrefix(sql, "/*") {
```
Comment (qiuyesuifeng) [inline tidb.go]: How about strings.HasPrefix(sql, "/*") is false, then comment will not be trimed?

File: tidb.go (hunk 1 above)
Comment (siddontang) [inline tidb.go]: what I mean is to check the trimming result, not only IsQuery true or false.

Comment (shenli) [inline tidb.go]: @ngaut We can handle the sql in #199 now. For sql text start with --, we just need to drop the entire sql text and do nothing else.

➜  interpreter git:(shenli/comment-in-query) ✗ cat xx
 /* mysql-connector-java-5.1.35 ( Revision: 5fb9c5849535c13917c2cf9baaece6ef9693ef27 ) */SHOW VARIABLES WHERE Variable_name ='language' OR Variable_name = 'net_write_timeout' OR Variable_name = 'interactive_timeout' OR Variable_name = 'wait_timeout' OR Variable_name = 'character_set_client' OR Variable_name = 'character_set_connection' OR Variable_name = 'character_set' OR Variable_name = 'character_set_server' OR Variable_name = 'tx_isolation' OR Variable_name = 'transaction_isolation' OR Variable_name = 'character_set_results' OR Variable_name = 'timezone' OR Variable_name = 'time_zone' OR Variable_name = 'system_time_zone' OR Variable_name = 'lower_case_table_names' OR Variable_name = 'max_allowed_packet' OR Variable_name = 'net_buffer_length' OR Variable_name = 'sql_mode' OR Variable_name = 'query_cache_type' OR Variable_name = 'query_cache_size' OR Variable_name = 'license' OR Variable_name = 'init_connect';
➜  interpreter git:(shenli/comment-in-query) ✗ cat xx | ./interpreter 
Welcome to the TiDB.
Version:
Git Commit Hash: None
UTC Build Time:  None

+--------------------------+------------------------+
| Variable_name            | Value                  |
+--------------------------+------------------------+
| character_set_results    | latin1                 |
| init_connect             |                        |
| character_set_client     | latin1                 |
| tx_isolation             | REPEATABLE-READ        |
| interactive_timeout      | 28800                  |
| max_allowed_packet       | 4194304                |
| lower_case_table_names   | 2                      |
| character_set_connection | latin1                 |
| time_zone                | SYSTEM                 |
| sql_mode                 | NO_ENGINE_SUBSTITUTION |
| query_cache_type         | OFF                    |
| net_write_timeout        | 60                     |
| system_time_zone         | CST                    |
| query_cache_size         | 1048576                |
| net_buffer_length        | 16384                  |
| license                  | GPL                    |
| wait_timeout             | 28800                  |
| character_set_server     | latin1                 |
+--------------------------+------------------------+

Bye

File: tidb.go (hunk 4 above)
Comment (siddontang) [inline tidb.go]: Have same doubt too.

Comment (shenli) [inline tidb.go]: This function is used to fix bug when sql text start with "/*comment*/" IsQuery always returns false.
We do not need to care about other comment format.

File: tidb.go (hunk 1 above)
Comment (qiuyesuifeng) [inline tidb.go]: Yes, it is my fault.
SQL like `select * /* comment */ from t;` is supported now.
Seems we only not support start with '/*' sql, like `/* comment */ select * from t`

Comment (ngaut) [inline tidb.go]: Looks like mp didn't handle comments start -- correctly

Comment (shenli) [inline tidb.go]: I test tidb-server and it is ok now.

Database changed
mysql> select /*comment*/ 1;
+------+
| 1    |
+------+
|    1 |
+------+
1 row in set (0.00 sec)

mysql> /*comment*/ select 1;
+------+
| 1    |
+------+
|    1 |
+------+
1 row in set (0.00 sec)

mysql>

File: tidb.go [hunk 5]
```diff
@@ -243,9 +243,24 @@ func NewStore(uri string) (kv.Storage, error) {
...
+		i := strings.Index(sql, "*/")
+		if i != -1 && i < len(sql)+1 {
+			sql = sql[i+2:]
+		}
```
Comment (qiuyesuifeng) [inline tidb.go]: If SQL like `/* select * from t;`, here seems will be an endless loop？

Comment (shenli) [inline tidb.go]: PTAL

File: parser/parser_test.go [hunk 6]
```diff
@@ -335,19 +335,21 @@ func (s *testParserSuite) TestParser0(c *C) {
...
 		ok := yyParse(l) == 0
 		c.Assert(ok, Equals, t.ok, Commentf("source %v", t.src))
 
 		switch ok {
 		case true:
-			c.Assert(len(l.errs), Equals, 0)
+			c.Assert(l.errs, HasLen, 0, Commentf("src: %s", t.src))
 		case false:
-			c.Assert(len(l.errs), Not(Equals), 0)
+			c.Assert(len(l.errs), Not(Equals), 0, Commentf("src: %s", t.src))
```
Comment (qiuyesuifeng) [inline parser/parser_test.go]: Not(HasLen)?

Comment (shenli) [inline parser/parser_test.go]: In the next pr.

**Comment by qiuyesuifeng** [comment]
LGTM

**Comment by shenli** [comment]
PTAL @ngaut @siddontang

**Comment by siddontang** [comment]
LGTM

**Pull request merged**